### PR TITLE
MEN-3411 Inventory client backport

### PR DESCRIPTION
**Pull request by merlin-northern**
ChangeLog:none
Signed-off-by: Peter Grzybowski <peter@northern.tech>

### CODE REVIEW

**Comment by merlin-northern** [comment]
Needed for the MEN-3411, I wanted to separate this backport from the implementation itself to stay as clean as possible.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ include:
     file: '.gitlab-ci-github-status-updates.yml'
 
 test:unit:
-  image: golang:1.11
+  image: golang:1.13
```
Comment (tranchitella) [inline .gitlab-ci.yml:25]: I believe golang:1.13 is the default (see https://github.com/mendersoftware/mendertesting/blob/master/.gitlab-ci-check-golang-unittests.yml#L30), maybe it makes sense to stick to the default?

File: Dockerfile.acceptance-testing [hunk 2]
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.11 as builder
+FROM golang:1.13 as builder
```
Comment (lluiscampos) [inline Dockerfile.acceptance-testing:1]: Thanks for the update. The main `Dockerfile` was already using 1.13, so now tests and binary are using the same :+1:

File: main_test.go [hunk 3]
```diff
@@ -36,10 +36,10 @@ func init() {
 
 	flag.BoolVar(&runAcceptanceTests, "acceptance-tests", false, "set flag when running acceptance tests")
 	flag.StringVar(&cliArgsRaw, "cli-args", "", "for passing urfave/cli args (single string) when golang flags are specified (avoids conflict)")
-	flag.Parse()
 }
 
 func TestRunMain(t *testing.T) {
+	flag.Parse()
```
Comment (oleorhagen) [inline main_test.go:42]: Why?

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,20 @@
+module github.com/mendersoftware/deployments
+
+go 1.13
+
+require (
+	github.com/ant0ine/go-json-rest v3.3.3-0.20170913041208-ebb33769ae01+incompatible
+	github.com/asaskevich/govalidator v0.0.0-20180720115003-f9ffefc3facf
```
Comment (oleorhagen) [inline go.mod:7]: Question:
Why is this on such an old sha? And not a released version?

File: main_test.go (hunk 3 above)
Comment (merlin-northern) [reply main_test.go:42]: part of the migration to go v.1.13, required.
e.g.: https://golang.org/src/flag/example_test.go
_"// execute, typically at the start of main (not init!):"_

File: go.mod (hunk 4 above)
Comment (merlin-northern) [reply go.mod:7]: believe it or not, but `TestUnmarshalLogMessage` fails with the latest. 
* github.com/asaskevich/govalidator v0.0.0-20200428143746-21a406dcc535 test fails
* github.com/asaskevich/govalidator v0.0.0-20180720115003-f9ffefc3facf test passes
just change in the dep made the test fail, it was not that easy to find.

File: model/device.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	//device's group name
+	Group string `json:"-" bson:"group,omitempty"`
+
+	CreatedTs time.Time `json:"-" bson:"created_ts,omitempty"`
+	//Timestamp of the last attribute update.
+	UpdatedTs time.Time `json:"updated_ts" bson:"updated_ts,omitempty"`
+}
+
+// wrapper for device attributes names and values
+type DeviceAttributes map[string]DeviceAttribute
```
Comment (alfrunes) [inline model/device.go:61]: Why do you need this? Doesn't it suffice to use `[]DeviceAttribute` ?

File: model/device.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	Value       interface{} `json:"value" bson:",omitempty"`
+	Scope       string      `json:"scope" bson:",omitempty"`
+}
+
+// Inventory device wrapper
+type InvDevice struct {
+	//system-generated device ID
+	ID string `json:"id" bson:"_id,omitempty"`
+
+	//a map of attributes names and their values.
```
Comment (alfrunes) [inline model/device.go:33]: This comment doesn't apply.

File: model/device.go (hunk 5 above)
Comment (merlin-northern) [reply model/device.go:61]: it is just copied from the deployments-enterprise, shall I remove it?

File: model/device.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	//device's group name
+	Group string `json:"-" bson:"group,omitempty"`
+
+	CreatedTs time.Time `json:"-" bson:"created_ts,omitempty"`
+	//Timestamp of the last attribute update.
+	UpdatedTs time.Time `json:"updated_ts" bson:"updated_ts,omitempty"`
+}
+
+// Device wrapper
+type Device struct {
```
Comment (alfrunes) [inline model/device.go:45]: I can't see this definition being used anywhere.

File: model/device.go (hunk 6 above)
Comment (merlin-northern) [reply model/device.go:33]: yeah, cool, but it still is from the ent counterpart. maybe lets leave it for the sake of next re-back-port-merge to enterprise?

File: model/device.go (hunk 7 above)
Comment (merlin-northern) [reply model/device.go:45]: I understand, but keeping it here may make one less conflict with the merge to ent. do you want me to remove it?

File: model/device.go (hunk 5 above)
Comment (alfrunes) [reply model/device.go:61]: I think you can remove this and the Device definition (if it's not being used anywhere), but maybe do it in a follow-up PR if this applies to the enterprise counterpart as well.

File: model/device.go (hunk 7 above)
Comment (alfrunes) [reply model/device.go:45]: Keep it in, but I'd suggest to remove it in a follow up PR and propagate to enterprise as well.

**Comment by merlin-northern** [comment]
> @merlin-northern ahh, I hit approve but didn't notice that vendor is gone, thought it was just an update.
> I was certain we decided to keep it, so not sure how to proceed - is it on purpose, am I missing something ? :)

I removed ./vendor, and recreated it with `go mod vendor` resulting is this commit:
https://github.com/mendersoftware/deployments/pull/437/commits/ce5f5b34afaebfd68294001e6a2102e376c0ee76
I guess you can call it heavily-trimmed vendor tree, since it now contains 1341 files, while the old one contained 1662. if you have objections we can postpone the migration to modules.

**Comment by mchalski** [comment]
@merlin-northern sorry for the confusion, I see it now.

File: model/device.go (hunk 5 above)
Comment (merlin-northern) [reply model/device.go:61]: https://github.com/mendersoftware/deployments/pull/439

File: model/device.go (hunk 7 above)
Comment (merlin-northern) [reply model/device.go:45]: https://github.com/mendersoftware/deployments/pull/439

**Pull request merged**